src/lightning/pytorch/utilities/restricted_classmethod.py

```diff
@@ -0,0 +1,23 @@
+from typing import Callable, Concatenate, Generic, ParamSpec, TypeVar
+
+_T = TypeVar("_T")  # type of the method owner
+_P = ParamSpec("_P")  # parameters of the decorated method
+_R_co = TypeVar("_R_co", covariant=True)  # return type of the decorated method
+
+
+class restricted_classmethod(Generic[_T, _P, _R_co]):
```

Comment:
I suggest to make this protected so we can change it in the future if needed. 

```suggestion
class _restricted_classmethod(Generic[_T, _P, _R_co]):
```